### ISSUE TITLE
Bootstrap failure with system absl

### ISSUE THREAD

**Issue by leahneukirchen** [just-buildsystem/justbuild#39]
Filing this issue because it's a regression from 1.1.4 and I can't debug it further atm.

Building justbuild 1.2.0 with NON_LOCAL_DEPS *not* containing `com_google_absl` now fails:

```
INFO: Requested target is [["@","just","",""],{"ADD_CFLAGS":["-Wno-error"],"ADD_CXXFLAGS":["-Wno-error"],"AR":"/usr/bin/ar","ARCH":"x86_64","CC":"cc","CFLAGS":["-fstack-clash-protection","-D_FORTIFY_SOURCE=2","-mtune=generic","-O2","-pipe","-ffile-prefix-map=/builddir/justbuild-1.2.0=."],"COMPILER_FAMILY":"gnu","CXX":"g++","CXXFLAGS":["-fstack-clash-protection","-D_FORTIFY_SOURCE=2","-mtune=generic","-O2","-pipe","-ffile-prefix-map=/builddir/justbuild-1.2.0=."],"ENV":{"PKG_CONFIG_PATH":"/usr/lib/pkgconfig:/usr/share/pkgconfig"},"LDFLAGS":["-Wl,-z,relro","-Wl,-z,now","-Wl,--as-needed"],"OS":"linux","SOURCE_DATE_EPOCH":1693061128.0,"VERSION_EXTRA_SUFFIX":"-void"}]
ERROR: While processing targets:
       While evaluating export target ["@","just","",""]:
       While analysing ["@","rules-just","CC","install-with-deps"] target ["@","just","","installed just"]:
       While evaluating configure target ["@","just","","just"]:
       While evaluating export target ["@","just","","exported-just"]:
       While analysing ["@","rules-just","CC","binary"] target ["@","just","src/buildtool/main","just"]:
       While analysing ["@","rules-just","CC","library"] target ["@","just","src/buildtool/storage","storage"]:
       While analysing ["@","rules-just","CC","library"] target ["@","just","src/buildtool/common","bazel_types"]:
       While analysing ["@","rules-just","CC/proto","service library"] target ["#","a0d338d0250537ba57c641b3c933e548bdd24d9b","9476599004d9f377bcc71f081af6c0ca4bac0d85"]:
       While analysing ["@","rules-just","CC/proto","library"] target ["#","a0d338d0250537ba57c641b3c933e548bdd24d9b","2c640c9332e66c97a48882cd3211387624e193e9"]:
       While analysing ["@","rules-just","CC/proto","defaults"] target ["@","rules-just","CC/proto","defaults"]:
       While evaluating export target ["@","protobuf","","libprotobuf"]:
       While analysing ["@","rules-protobuf","CC","library"] target ["@","protobuf","src/google/protobuf","libprotobuf"]:
       While searching targets description for ["@","com_google_absl","absl/strings","internal"]:
       JSON file absl/strings/TARGETS.absl does not exist.
```

and more errors for the other `absl/*` directories.

System absl is version 20230802.0.  This works fine with justbuild 1.1.4.

(I wouldn't mind vendoring absl, but then it fails on x86_64-musl due to some legacy `off64_t` usage...)

**Comment by aehlig** [comment]
Thanks for reporting this issue. This is indeed a use case we never had in mind. As `justbuild` does not depend directly on `absl`, but only indirectly through `protobuf` (and `grpc`), we never added target files for absl in the `etc/import.pkconfig` assuming that all depdencies on `absl` were picked up through the `pkg-config` files of the direct dependencies. With the update of the bundled `protobuf` version, however, the bundled `protobuf` now newly depends on `absl`. I guess we have to add target files describing how to pick up the individual `absl` libraries via `pkg-config`.

**Comment by aehlig** [comment]
An alternative, abeit a bit hacky, work around might be to vendor `absl` and add in the [CC defaults for `absl`](https://github.com/just-buildsystem/justbuild/blob/a4a59ea849cc439ee67665d0668d8e1f71dfa72e/etc/defaults/CC/TARGETS.absl#L23) a suitable `-Doffset64_t=...` flag.

**Comment by leahneukirchen** [comment]
Aha, the issue is fixed in 20230802.0 but exists in 20230125.3, which is the vendored version... perhaps I can just bump that.

**Comment by leahneukirchen** [comment]
Of course, nothing is easy as 20230802.0 shuffles paths around.

As an aside, how does one easily update an repository/type "archive"?

**Comment by aehlig** [comment]
> Of course, nothing is easy as 20230802.0 shuffles paths around.

On the other hand, absl 20230125.3 contains a [single occurence of `off64_t`](https://github.com/abseil/abseil-cpp/blob/c2435f8342c2d0ed8101cb43adfd605fdc52dca2/absl/base/internal/direct_mmap.h#L75). So, if this is the only issue, we can as well logically patch that file if we do not want to go the way of [preprocessing it away](https://github.com/just-buildsystem/justbuild/issues/39#issuecomment-1694649085).

> As an aside, how does one easily update an repository/type "archive"?

We have not yet built any automation for this, so, at the moment, you have to update by hand
- the url to fetch,
- the `"content"` field (containing the git blob identifiers), and
- (likely) the subdirectory to take, as well as
- any additional (besides the git blob identifier) checksums (`"sha256"`, `"sha512"`) you want to be verified after download.

**Comment by leahneukirchen** [comment]
All of these are easy to update, except for the content field, which requires me digging into the store (unless I'm missing something).

**Comment by aehlig** [comment]
> All of these are easy to update, except for the content field, which requires me digging into the store (unless I'm missing something).

You can also use `git hash-object`. E.g.,

```
/tmp>git hash-object 20230125.3.tar.gz
08a0c4a58d13f234bd3f1de6c73395ec5ab7d5f6
/tmp>
```

**Comment by aehlig** [comment]
> On the other hand, absl 20230125.3 contains a [single occurence of `off64_t`](https://github.com/abseil/abseil-cpp/blob/c2435f8342c2d0ed8101cb43adfd605fdc52dca2/absl/base/internal/direct_mmap.h#L75). So, if this is the only issue, we can as well logically patch that file if we do not want to go the way of [preprocessing it away](https://github.com/just-buildsystem/justbuild/issues/39#issuecomment-1694649085).

Such a logical patching would look something like https://github.com/aehlig/justbuild/commit/1b068c12b31b4a484bf49964b3057e6466b8953a Please let us know if that fixes the issue.

**Comment by leahneukirchen** [comment]
That fixes it, yes.

**Comment by aehlig** [comment]
> That fixes it, yes.

As these changes are now in a released version, [release 1.2.1](https://github.com/just-buildsystem/justbuild/releases/tag/v1.2.1), I'm closing the issue
